feat: enrich configuration override rule (#1811)
emqtt/emq-docs-cn: en_US/configuration/configuration.md

### Configuration Rewrite File

> `emqx.conf` defines settings at a global level, for cases where you need to customize the settings for a cluster or a node, EMQX also provides a configuration rewrite file implicitly under `emqx.conf`:

Reviewer: ```suggestion
`emqx.conf` defines settings at a global level, for cases where you need to customize the settings for a cluster or a node, EMQX also provides a configuration rewrite file which extends but does not override `emqx.conf`:
```

It is possible to change data directory from config `node.data_dir` or environment variable `EMQX_NODE__DATA_DIR`, however, when running a cluster, all nodes should have the same path.
:::

> By default, most global settings are defined in the `emqx.conf` file, if you perform certain operations on the cluster level with Dashboard, REST API or CLI, the changes will be synced with the `cluster.hocon` as configuration files and overide the corresponding settings in `emqx.conf`. And this whole process is called hot reload.

Reviewer: ```suggestion
By default, most global settings are defined in the `emqx.conf` file, if you perform certain operations on the cluster level from Dashboard, REST API or CLI, the changes will be stored in `cluster.hocon`. And this whole process is called hot reload.
```

- In the same file, the value defined in the later section will override any previous key value.
- A higher-level value will replace that of a lower-level.

the priority of configuration is Environment Variables, `etc/emqx.conf` file and HTTP API.
This means that for a newly started EMQX, the configuration settings in environment variables and
`etc/emqx.conf` will always override any configurations made in HTTP API.

If the configuration is modified through the HTTP API while EMQX is running, it will take effect immediately in the cluster. However, if EMQX is restarted, these modified configurations may be overwritten by the Environment Variables or etc/emqx.conf

Configurations modified via the HTTP API will take effect during EMQX running. However, after

> restarting EMQX, they may be overwritten by higher priority Environment Variables or `etc/emqx.conf`, resulting in lost these configurations. Therefore, the best approach is to manually persist all configurations modified via HTTP API to the `etc/emqx.conf` file to prevent them lost.

Author: The order of priority for configuration is:
environment variables, 'etc/emqx.conf', and 'data/configs/cluster.hocon'.
1. Settings in environment variables starting with 'EMQX_' are given the highest priority.
2. Configuration settings in the 'etc/emqx.conf' file may be overwritten by environment variables.
3. Changes made through the EMQX dashboard UI, management HTTP API, or CLI
   overwrite the 'data/configs/cluster.hocon' file at runtime.
 4. If you modify a configuration setting using the API, the change takes effect immediately.
 5. However, if the same setting is configured differently in the 'etc/emqx.conf' file,
   'etc/emqx.conf' takes priority after a reboot.

Author: which one is better?

- In the same file, the value defined in the later section will override any previous key value.
- A higher-level value will replace that of a lower-level.

The EMQX configuration is prioritized (overlayed) in the following order: `cluster.hocon < emqx.conf < environment variables`.

Settings in environment variables that begin with 'EMQX_' have the highest priority and will override any settings in the `etc/emqx.conf` file.

> Changes made through the EMQX dashboard UI, management HTTP API, or CLI will be written into the `data/configs/cluster.hocon` file at runtime and will take effect immediately.
> 
> However, if the same configuration items are set differently in the `etc/emqx.conf` file, the runtime updates will be overridden by the settings in `etc/emqx.conf` after the node restarts.

Reviewer: ```suggestion
Changes made through EMQX Dashboard UI, HTTP API, or CLI are persisted in `data/configs/cluster.hocon` at runtime and will take effect immediately. However, if the same configuration items are set differently in the `etc/emqx.conf` file, the runtime updates will be overridden by the settings in `etc/emqx.conf` after the node restarts.
```

To avoid confusion, it is highly recommend NOT to have the same config keys in both `cluster.hocon` and `emqx.conf`.

::: tip

> 1. If you're using an older version of EMQX, specifically version e5.0.3/v5.0.22 or earlier(i.e. the `cluster-override.conf` file still exists in EMQX's data directory),

Reviewer: ```suggestion
1. If you're using an older version of EMQX, specifically version e5.0.2/v5.0.22 or earlier(i.e. the `cluster-override.conf` file still exists in EMQX's data directory),
```

   then the order of priority for configuring your settings is as follows: `emqx.conf < ENV < HTTP API(cluster-override.conf)`.

> 2. If you're upgrading from e5.0.3/v5.0.22 or earlier to the latest version of EMQX, 

Reviewer: ```suggestion
2. If you're upgrading from e5.0.2/v5.0.22 or earlier to the latest version of EMQX, 
```

>    the configuration priority will remain consistent with the previous version in order to maintain compatibility.  

Reviewer: ```suggestion
   the configuration overriding order will remain unchanged, `cluster.hocon` will not be created to keep compatibility.  
```

- In the same file, the value defined in the later section will override any previous key value.
- A higher-level value will replace that of a lower-level.

> The EMQX configuration is prioritized in the following order: `environment variables > emqx.conf > API(cluster.hocon)`.
> ```
>                     Start
>                       |
>             Check Environment Variables
>                  /          \
>              Found         Not Found
>               |                  |
>            Use Env        Check emqx.conf
>                            /             \
>                         Found           Not Found
>                           |                  |
>                     Use emqx.conf    Check cluster.hocon
>                                           /          \
>                             Use cluster.hocon      Use Default Values
> 
> ```

Reviewer: ```suggestion
The EMQX configuration is prioritized (overlayed) in the following order: `cluster.hocon < emqx.conf < environment variables`.
```

Settings in environment variables that begin with 'EMQX_' have the highest priority and will override any settings in the `etc/emqx.conf` file.

Changes made through the EMQX dashboard UI, management HTTP API, or CLI will be written into the `data/configs/cluster.hocon` file at runtime and will take effect immediately.

> However, if the same configuration items are set differently in the `etc/emqx.conf` file, the runtime updates will be overridden by the settings in `etc/emqx.conf` during a restart.

Reviewer: ```suggestion
However, if the same configuration items are set differently in the `etc/emqx.conf` file, the runtime updates will be overridden by the settings in `etc/emqx.conf` after the node restarts.

To avoid confusion, it is highly recommend NOT to have the same config keys in both `cluster.conf` and `emqx.conf`.
```
